### PR TITLE
Optimize variant interface implementation generation

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Saves 106 kB on a Hello World app.

Previously when the compiler saw we're generating a type that implements variant interfaces, we would throw hands in the air and generate the implementations of all variant interface methods on the type since we had no idea which ones would be actually needed.

This has drawbacks, especially around array types that implement a lot of variant (or variant-like) interfaces.

This commit introduces tracking of individual variant interface methods being used. It lets us avoid generating implementations of variant interface methods that are not statically seen as necessary.

We track this by adding conditional dependencies on a new `VariantInterfaceMethodUseNode` instead of the old logic that just generated hard `VirtualMethodUse` dependencies.

We try to further optimize when a virtual method call or interface method implementation is actually considered variant. E.g. an implementation or a call to `IEnumerable<MyStruct>` is not actually variant since we can't cast it to any other `IEnumerable<T>`.

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/VariantInterfaceMethodUseNode.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            _decl = decl;
+        }
+
+        public static bool IsVariantInterfaceImplementation(NodeFactory factory, TypeDesc providingType, DefType implementedInterface)
+        {
+            Debug.Assert(implementedInterface.IsInterface);
+            Debug.Assert(!implementedInterface.IsGenericDefinition);
+
+            // If any of the implemented interfaces have variance, calls against compatible interface methods
+            // could result in interface methods of this type being used (e.g. IEnumberable<object>.GetEnumerator()
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/VariantInterfaceMethodUseNode.cs:53]:
```suggestion
            // could result in interface methods of this type being used (e.g. IEnumerable<object>.GetEnumerator()
```

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs [hunk 2]
```diff
@@ -284,6 +286,16 @@ public sealed override IEnumerable<CombinedDependencyListEntry> GetConditionalSt
                         if (implMethod != null)
                         {
                             yield return new CombinedDependencyListEntry(factory.VirtualMethodUse(implMethod), factory.VirtualMethodUse(interfaceMethod), "Interface method");
+
+                            // If any of the implemented interfaces have variance, calls against compatible interface methods
+                            // could result in interface methods of this type being used (e.g. IEnumberable<object>.GetEnumerator()
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:291]:
```suggestion
                            // could result in interface methods of this type being used (e.g. IEnumerable<object>.GetEnumerator()
```

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
Nice!

**Comment by MichalStrehovsky** [comment]
Sigh, the Pri-0 runs have a lot of failure. I bet I messed up the IL rewriter again. I'll take a look tomorrow.

**Pull request merged**